### ISSUE TITLE
ArchLinux LTS kernel is now the newer longterm v6.1 kernel

### ISSUE THREAD

**Issue by WolfWings** [linux-surface/linux-surface#1083]
[https://archlinux.org/packages/?name=linux-lts](https://archlinux.org/packages/?name=linux-lts)

I glanced at the PKGBUILD but it's a large enough skip from the existing -lts PKGBUILD that I didn't feel comfortable just tossing a PR in blindly and trying to get it right immediately since the LTS package pulls from kernel.org instead of the archlinux repos, different patches, etc. So... an issue that will likely be extremely quick to handle instead.

**Comment by qzed** [comment]
I'll have to think about reviving it. I guess now would be the best time to do that.

We've discontinued LTS kernels a while ago as the major point of their existence was the IPTS driver with the original touch processing firmware. This can't be implemented on newer kernels (4.19 was the last LTS kernel that worked) and it's gone for good now. So I'm not sure if there are any major benefits to keeping the LTS kernel around anymore. But feel free to try to convince me otherwise.

**Comment by WolfWings** [comment]
Ah, was unaware of that aspect and thought it existed since stock Arch has the 'mainline' kernel and 'lts' kernel packages so it was keeping distro parity not about the IPTS stuff. The huge number of kernel versions in the patches/ subdirectory made me think they were all maintained, but I'd missed the timestamps on most of them being quite old.

I guess it depends if you plan to keep patches for other LTS kernel versions updated or only target mainline/stable?

If older kernel versions won't be targeted should the patches be left in the repo trunk or removed entirely perhaps? Folks can just go into the older commit history if they need that older version patch.

**Comment by qzed** [comment]
Yeah, at the moment we only maintain the latest version. I think there was a request some time ago to keep the old patches around, but we should make it more clear as to what's maintained and what not. There's a README in the patches directory but I somewhat forgot to update that...

The biggest problem I have with LTS and older kernels is the additional maintenance burden. The difference between 6.1 and 6.2 isn't big (as far as our stuff is concerned), but things get more difficult once this difference gets larger and when new patches depend on new features, which would mean we have to maintain two largely different patchsets, with potential for different bugs... I think at the moment, we're in a state where this will probably be less of an issue than in the past though, since things have been largely stable. Another issue is that this likely leads to more reports of things that have been fixed upstream in newer versions. In particular if you run new devices with new chipsets... LTS will always be second-tier support.

But I have been thinking about trying to go for distro parity in kernel sources/versions in Ubuntu/Debian (for the others we essentially have that already) since there can be some problems with newer kernels on those distros. For that, we'd ideally make things work with their build system to ensure we also include their patches and configs, which didn't go that well last time I tried.

I also thought about automating rebasing and patch generation, but haven't found any solution that really works well and integrates nicely into GitHub actions. That would remove some of the more repetitive stuff, but it's also not _that_ time consuming.

**Comment by WolfWings** [comment]
Yeah the maintenance burden grows exponentially since you have to worry about when internal calls change from versions, etc, and juggling low-level hardware components makes that ten times worse, so to speak.

Even requests to keep older patches around... since any scripts can simply clone from a tag or commit ID just adding annotated tags (instead of simple ones that are just symlinks to a commit ID) for kernel versions solves that. Annotated tags can be re-pointed to a later commit if for some reason those patches get updated down the road.

For Arch I admit it's a special case since their LTS is always only the 'latest' LTS kernel, it's not like Debian/Ubuntu/Fedora which are almost all back on 5.10 or 5.15 still, so it won't fall as far behind.

Right now it seems like it's a bit of a mish-mash of indications of LTS support versus reality of... well... not? I admit my instinct is purge what's not actually 'valid' currently from trunk, but I admit I rely on that "It's never actually deleted." nature of git quite heavily compared to some developers I know. Better to remove something when it's unmaintained so folks have to go to the 'last maintained' version to get it, then re-add it again if it becomes maintained again.

If parity with mainstream distros ends up being taken that could reduce the list of kernel versions in /patches to:

 * 5.10 (lots)
 * 5.15 (also lots)
 * 6.1 (Arch and others in the coming months)
 * Mainline (6.2 as of this comment)

That would cover the latest releases that are less than 50% through their expected lifespan (basically excludes Ubuntu 20.04 which is the last bastion of 5.4), so that's still a severe slimming down and better indicates what's actually covered/checked on better at least?

**Comment by qzed** [comment]
> Right now it seems like it's a bit of a mish-mash of indications of LTS support versus reality of... well... not? I admit my instinct is purge what's not actually 'valid' currently from trunk, but I admit I rely on that "It's never actually deleted." nature of git quite heavily compared to some developers I know. Better to remove something when it's unmaintained so folks have to go to the 'last maintained' version to get it, then re-add it again if it becomes maintained again.

Yeah, I think you're right about that. We can just point people to the respective git kernel branches and tell them to get their patches from there (e.g. [`v5.4-surface`](https://github.com/linux-surface/kernel/tree/v5.4-surface)). Probably adding patchlevel tags would also make sense. The benefit of that is that it's a full kernel tree, so it should be guaranteed that it compiles (as opposed to applying some outdated patchset onto a newer kernel). I'll remove the LTS package scripts and the outdated patches for now. We can always add those back when we update/revive them.

I'll have a look at the kernel versions you mentioned. Those seem mostly sensible. I definitely wouldn't want to go back further though. Even 5.10 has a bunch of stuff that I'll likely need to sort out as `platform/surface` didn't exist back then (our 5.10 patches have that stuff in `patform/x86`, which makes backporting new changes to `platform/surface` a bit annoying). Hopefully, we can just backport the root commits of that and everything will work fine. It will take me some time to go through all the other patches though and see if that's feasible.

The remaining question is then how long do we support those LTS releases? I guess if we stick to those versions for now, we could say the last 3 LTS releases?

**Comment by WolfWings** [comment]
Either the 'last 3' or target distro's for parity + latest? I'm going to pull data from the Steam hardware survey since anyone running Linux there is coming from a desktop perspective instead of intermixing servers so it seems a bit more valid for here, and since the Steam Deck is a single fixed OS type it can be easily ignored.

So that leaves:
 * Arch Linux
 * Ubuntu 22.04
 * Manjaro (derivative of Arch)
 * Linux Mint 21 (derivative of Ubuntu 22.04)
 * Pop! OS 22.04 (another Ubuntu 22.04 derivative)

Those account for roughly 2/3rds of Linux users that use Steam, adding in Debian Stable + Devuan Stable + RHEL9 covers a broad swath of additional distros to try to adapt for the Steam-gamer bias.

Among those the currently used kernels end up the above list I made.

Arch/Manjaro is "latest Longterm + mainline" so would add 6.1 for LTS.
Fedora is mainline.
Ubuntu 22.04 and all derivatives is 5.15 w/ the "Hard Ware Enablement" kernel likely to change to 6.1.
Debian/Devuan and all derivatives are both 5.10.
RHEL9/RockyOS9/CentOS-Stream is 5.14, because of course RH would go off the rails a bit further? So adding it in for comparison doesn't mean I suggest supporting it TBH.

So yeah, by and large "mainline + most recent 3 LTS" would cover all scenarios even if I go 'back in time' and do the above kernel version checks for most points in the last few years. A couple majority distro's everyone else re-bases on lynchpin to 1-2 LTS releases and then 1-2 more pop up past that as distro's EOL. And would roughly say "Hey, use a kernel from the last 3 YEARS please if you're trying to get bleeding-edge devices to work?" which is... beyond reasonable. Comically generous even.

**Comment by qzed** [comment]
I would kind of favor the "last 3" since its quite a bit easier for me than figuring out which distros/distro-versions to consider and which kernel each of them use again :-) As you've mentioned this should cover quite a bit, so I'd guess chances are low that we're missing out on some important distro with that policy.

> RH would go off the rails a bit further?

Yeah, I don't think we want to add that (also not sure why you'd want RHEL/... on a Surface to be honest). At the moment, I'm thinking Fedora (latest, has no LTS), Arch + Arch LTS, Debian/Ubuntu mainline + 3 lts (so currently 6.2, 6.1, 5.15, 5.10; or less if that turns out to be too much).

**Comment by WolfWings** [comment]
Yeah, I wasn't advocating for anything more complex, more "Even the most complex idea I could think of... is still covered by 'last three' for the last several years so might as well keep it simple." And reducing that occasionally when some major change happens in the kernel would still be reasonable.

**Comment by qzed** [comment]
Okay, so I tried setting up a 5.10 kernel but there are a couple of problems: There are some build errors on Fedora 37 due to what seem to be deprecated functions and changed compiler defaults. So I wasn't able to properly compile and test it as a package for my system (it's probably possible to figure this out, but I'm not sure it's worth the time at the moment). I also I haven't been able to get the cameras working on v5.19 and earlier. Lastly, there's the problem that my SB2 is fairly well supported on v5.15 already and I can't test a large portion of the patches I'd have to backport.

So at this point that's a bit too much work for me right now. We can still aim for 3 LTS versions with the release of upcoming ones, but for now I'd stick to 6.1 as the only LTS. It's probably also quite a bit easier to just update that as we go instead of trying to backport a ton of history in one step.
